### PR TITLE
NFT Allowances REST API

### PR DESCRIPTION
**Pull request by mgoelswirlds**
**Description**:

This HIP designs Nft Allowances APIs for mirror node.

### CODE REVIEW

File: HIP/hip-857.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+      "spender": "0.0.9857",
+      "token_id": "0.0.1032",
+      "timestamp": {
+        "from": "1633466229.96884615",
+        "to": null
+      }
+    }
+  ],
+  "links": {
+		"next": "/api/v1/accounts/0.0.1000/allowances/nfts?limit=3&order=asc&account.id=gte:9857&token.id=gt:0.0.1034"
```
Comment (steven-sheehy) [inline HIP/hip-857.md:94]: limit=3 would have to be in the initial URL for this link to be accurate. Same for other example.

Comment (steven-sheehy) [inline HIP/hip-857.md:94]: We sort first by spender so we would exclude rows if gt:0.0.1034
```suggestion
		"next": "/api/v1/accounts/0.0.1000/allowances/nfts?limit=3&order=asc&account.id=gte:9857&token.id=gt:0.0.1032"
```

File: HIP/hip-857.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+  "links": {
+		"next": "/api/v1/accounts/0.0.8488/allowances/nfts?limit=3&order=asc&account.id=gte:1001&token.id=gt:0.0.1099"
+	}
+}
+```
+
+Query Parameters:
+
+- `account.id`: Filter by the spender account ID or owner account ID, depending on the owner flag. `ne` operator is not supported.
+- `limit`: The maximum number of items to return.
```
Comment (steven-sheehy) [inline HIP/hip-857.md:144]:
```suggestion
- `limit`: The maximum number of items to return. Defaults to 25 with a maximum of 100 allowed.
```

File: HIP/hip-857.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+- `order`: Order by `token.id` and `account.id`. Accepts `asc` or `desc` with a default of `asc`.
```
Comment (steven-sheehy) [inline HIP/hip-857.md:145]:
```suggestion
- `order`: Order by  `account.id` then `token.id`. Accepts `asc` or `desc` with a default of `asc`.
```

File: HIP/hip-857.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+- `owner`: Indicates whether the path parameter `accountId` is the owner or the spender ID. Accepts a boolean value of `true` or `false` with a default value set to `true`.
+- `token.id`: Filter by the token ID. `ne` operator is not supported.
```
Comment (steven-sheehy) [inline HIP/hip-857.md:147]:
```suggestion
- `token.id`: Filter by the token ID. `ne` operator is not supported. Only one occurrence allowed.
```

File: HIP/hip-857.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+  ],
+  "links": {
+		"next": "/api/v1/accounts/0.0.8488/allowances/nfts?limit=3&order=asc&account.id=gte:1001&token.id=gt:0.0.1099"
+	}
+}
+```
+
+Query Parameters:
+
+- `account.id`: Filter by the spender account ID or owner account ID, depending on the owner flag. `ne` operator is not supported.
```
Comment (steven-sheehy) [inline HIP/hip-857.md:143]:
```suggestion
- `account.id`: Filter by the spender account ID or owner account ID, depending on the owner flag. `ne` operator is not supported. Only one occurrence allowed.
```

File: HIP/hip-857.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+- `order`: Order by `token.id` and `account.id`. Accepts `asc` or `desc` with a default of `asc`.
+- `owner`: Indicates whether the path parameter `accountId` is the owner or the spender ID. Accepts a boolean value of `true` or `false` with a default value set to `true`.
+- `token.id`: Filter by the token ID. `ne` operator is not supported.
+
+Pagination is important to implement because there are accounts that could have a large number of NFT allowances, making a non-paginated response impractical. This requires multi-column pagination, including owner, spender, and token IDs.
+
+**Ordering**
+
+The order is governed by a combination of the account ID and the token ID values, with the account ID being the parent column. The token ID value governs its order within the given account ID.
+The default order for this API is currently ascending.
```
Comment (steven-sheehy) [inline HIP/hip-857.md:154]:
```suggestion
The default order for this API is ascending.
```

File: HIP/hip-857.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+The table below defines the restrictions and support for the endpoint.
+
+| Query Param | Comparison Operator | Support | Description | Example |
+| --- | --- | --- | --- | --- |
+| account.id | eq | Y | Single occurrence only. | ?account.id=X |
+|  | ne | N |  |  |
+|  | lt(e) | Y | Single occurrence only. | ?account.id=lte:X |
+|  | gt(e) | Y | Single occurrence only. | ?account.id=gte:X |
+| token.id | eq | Y | Single occurrence only. Requires the presence of an `account.id` query parameter| ?token.id=lt:Y |
```
Comment (steven-sheehy) [inline HIP/hip-857.md:168]:
```suggestion
| token.id | eq | Y | Single occurrence only. Requires the presence of an `account.id` query parameter| ?account.id=X&token.id=eq:Y |
```

File: HIP/hip-857.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+|  | lt(e) | Y | Single occurrence only. Requires the presence of a `lte` or `eq` `account.id` query parameter| ?account.id=lte:X&token.id=lt:Y |
+|  | gt(e) | Y | Single occurrence only. Requires the presence of a `gte` or `eq` `account.id` query parameter| ?account.id=gte:X&token.id=gt:Y |
+
+Both filters must be a single occurrence of **gt(e)** or **lt(e)** which provide a lower and or upper boundary for search.
+
+### Serial level allowance information
+
+To retrieve NFT serial number level allowance information with a `tokenId`, you can use an existing API.
+
+GET`api/v1/tokens/{tokenId}/listNfts`
```
Comment (steven-sheehy) [inline HIP/hip-857.md:179]:
```suggestion
GET `/api/v1/tokens/{tokenId}/nfts`
```

File: HIP/hip-857.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+| token.id | eq | Y | Single occurrence only. Requires the presence of an `account.id` query parameter| ?token.id=lt:Y |
+|  | ne | N |  |  |
+|  | lt(e) | Y | Single occurrence only. Requires the presence of a `lte` or `eq` `account.id` query parameter| ?account.id=lte:X&token.id=lt:Y |
+|  | gt(e) | Y | Single occurrence only. Requires the presence of a `gte` or `eq` `account.id` query parameter| ?account.id=gte:X&token.id=gt:Y |
+
+Both filters must be a single occurrence of **gt(e)** or **lt(e)** which provide a lower and or upper boundary for search.
+
+### Serial level allowance information
+
+To retrieve NFT serial number level allowance information with a `tokenId`, you can use an existing API.
```
Comment (steven-sheehy) [inline HIP/hip-857.md:177]:
```suggestion
To retrieve NFT serial number level allowance information for a specific token ID you can use the existing API:
```

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by Ashe-Oro — APPROVED** [review]
looks good! great feature and thank you both :)

**Pull request merged**